bindings/pyroot_experimental/PyROOT/src/PyzPythonHelpers.cxx

```diff
@@ -19,12 +19,116 @@ PyROOT extension module.
+{
+   if (!obj) {
+      PyErr_SetString(PyExc_RuntimeError, "Invalid Python object.");
+      return NULL;
+   }
+
+   // Get array interface of object
+   auto pyinterface = PyObject_GetAttrString(obj, "__array_interface__");
+   if (!pyinterface) {
+      PyErr_SetString(PyExc_RuntimeError, "__array_interface__ does not exist.");
+      return NULL;
+   }
+   if (!PyDict_Check(pyinterface)) {
+      PyErr_SetString(PyExc_RuntimeError, "__array_interface__ is not a dictionary.");
+      return NULL;
+   }
+
+   // Get the data-pointer
+   auto pydata = PyDict_GetItemString(pyinterface, "data");
+   if (!pydata) {
+      PyErr_SetString(PyExc_RuntimeError, "__array_interface__['data'] does not exist.");
+      return NULL;
+   }
+   long data = PyLong_AsLong(PyTuple_GetItem(pydata, 0));
+
+   // Get the size of the contiguous memory
+   auto pyshape = PyDict_GetItemString(pyinterface, "shape");
+   if (!pyshape) {
+      PyErr_SetString(PyExc_RuntimeError, "__array_interface__['shape'] does not exist.");
+      return NULL;
+   }
+   long size = 0;
+   for (unsigned int i = 0; i < PyTuple_Size(pyshape); i++) {
+      if (size == 0) size = 1;
+      size *= PyLong_AsLong(PyTuple_GetItem(pyshape, i));
+   }
+
+   // Get the typestring
+   auto pytypestr = PyDict_GetItemString(pyinterface, "typestr");
+   if (!pytypestr) {
+      PyErr_SetString(PyExc_RuntimeError, "__array_interface__['typestr'] does not exist.");
+      return NULL;
+   }
+   std::string typestr = CPyCppyy_PyUnicode_AsString(pytypestr);
+   const auto length = typestr.length();
+   if(length != 3) {
+      PyErr_SetString(PyExc_RuntimeError,
+              ("__array_interface__['typestr'] returned '" + typestr + "' with invalid length unequal 3.").c_str());
+      return NULL;
+   }
+
+   // Verify correct endianess
+   const auto endianess = typestr.substr(1, 2);
+#ifdef R__BYTESWAP
+   const auto byteswap = "<";
+#else
+   const auto byteswap = ">";
+#endif
+   if (!endianess.compare(byteswap)) {
+      PyErr_SetString(PyExc_RuntimeError, "Endianess of __array_interface__['typestr'] does not match endianess of ROOT.");
+      return NULL;
+   }
+
+   const auto dtype = typestr.substr(1, length);
+   std::string cppdtype;
+   if (dtype == "i4") {
+      cppdtype = "int";
+   } else if (dtype == "u4") {
+      cppdtype = "unsigned int";
+   } else if (dtype == "i8") {
+      cppdtype = "long";
+   } else if (dtype == "u8") {
+      cppdtype = "unsigned long";
+   } else if (dtype == "f4") {
+      cppdtype = "float";
+   } else if (dtype == "f8") {
+      cppdtype = "double";
+   } else {
+      PyErr_SetString(PyExc_RuntimeError, ("Python object has unknown data-type '" + dtype + "'.").c_str());
+      return NULL;
+   }
+
+   // Construct an RVec of the correct data-type
+   const std::string klassname = "ROOT::VecOps::RVec<" + cppdtype + ">";
+   auto address = (void*) gInterpreter->Calc(klassname + "(reinterpret_cast<" + cppdtype + "*>(" + data + ")," + size + ")");
```

Comment:
@Axel-Naumann @etejedor Is this the "safe way" to do so? Is the lifetime of the C++ object indefinite (until the garbage collection destroys it ...)

Reply:
Note to myself: Probably we need to check with the flags you can provide for `CPyCppyy::BindCppObject(address, klass, flags)` and make sure that the object is actually cleaned up.

Reply:
Yes it is until the interpreter cleans it.